File: app/src/main/java/jp/hazuki/yuzubrowser/speeddial/view/SpeedDialSettingActivityEditFragment.java
package jp.hazuki.yuzubrowser.speeddial.view;

import android.app.Activity;
import android.content.ActivityNotFoundException;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Bundle;
import android.provider.MediaStore;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CompoundButton;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.Switch;
import android.widget.Toast;

import java.io.File;

import jp.hazuki.yuzubrowser.R;
import jp.hazuki.yuzubrowser.RootLayout;
import jp.hazuki.yuzubrowser.download.DownloadFileProvider;
import jp.hazuki.yuzubrowser.speeddial.SpeedDial;
import jp.hazuki.yuzubrowser.speeddial.WebIcon;
import jp.hazuki.yuzubrowser.utils.ImageUtils;

public class SpeedDialSettingActivityEditFragment extends Fragment {

    private static final String DATA = "dat";
    private static final int REQUEST_PICK_IMAGE = 100;
    private static final int REQUEST_CROP_IMAGE = 101;

    private SpeedDial speedDial;
    private EditText name;
    private EditText url;
    private ImageButton icon;
    private View bottomBar;
    private SpeedDialEditCallBack mCallBack;
    private GoBackController goBack;

    public static Fragment newInstance(SpeedDial speedDial) {
        Fragment fragment = new SpeedDialSettingActivityEditFragment();
        Bundle bundle = new Bundle();
        bundle.putSerializable(DATA, speedDial);
        fragment.setArguments(bundle);
        return fragment;
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_edit_speeddial, container, false);

        speedDial = (SpeedDial) getArguments().getSerializable(DATA);

        if (speedDial == null) {
            speedDial = new SpeedDial();
        }

        name = (EditText) v.findViewById(R.id.name);
        url = (EditText) v.findViewById(R.id.url);
        icon = (ImageButton) v.findViewById(R.id.imageButton2);
        bottomBar = v.findViewById(R.id.bottomBar);

        Switch sw = (Switch) v.findViewById(R.id.use_favicon);

        ((RootLayout) v.findViewById(R.id.superFrameLayout)).setOnImeShownListener(new RootLayout.OnImeShownListener() {
            @Override
            public void onImeVisibilityChanged(boolean visible) {
                bottomBar.setVisibility(visible ? View.GONE : View.VISIBLE);
            }
        });

        name.setText(speedDial.getTitle());
        url.setText(speedDial.getUrl());

        if (speedDial.getIcon() == null) {
            speedDial.setIcon(WebIcon.createIcon(ImageUtils.getBitmapFromVectorDrawable(getActivity(), R.drawable.ic_public_white_24dp)));
        }

        icon.setImageBitmap(speedDial.getIcon().getBitmap());

        sw.setChecked(speedDial.isFavicon());
        setIconEnable(!speedDial.isFavicon());

        sw.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                speedDial.setFaviconMode(isChecked);
                setIconEnable(!isChecked);
            }
        });

        icon.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent();
                intent.setType("image/*");
                intent.setAction(Intent.ACTION_GET_CONTENT);
                Intent i = new Intent(Intent.ACTION_PICK, MediaStore.Images.Media.EXTERNAL_CONTENT_URI);
                startActivityForResult(i, REQUEST_PICK_IMAGE);
            }
        });

        v.findViewById(R.id.okButton).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (mCallBack != null) {
                    speedDial.setTitle(name.getText().toString());
                    speedDial.setUrl(url.getText().toString());
                    mCallBack.onEdited(speedDial);
                }
            }
        });

        v.findViewById(R.id.cancelButton).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (goBack != null)
                    goBack.goBack();
            }
        });

        return v;
    }

    private void setIconEnable(boolean enable) {
        icon.setEnabled(enable);
        icon.setAlpha(enable ? 1.0f : 0.6f);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        switch (requestCode) {
            case REQUEST_PICK_IMAGE:
                if (resultCode == Activity.RESULT_OK) {
                    try {
                        Uri uri = data.getData();
                        if ("file".equals(uri.getScheme())) {
                            uri = DownloadFileProvider.getUriForFIle(new File(uri.getPath()));
                        }
                        Intent intent = new Intent("com.android.camera.action.CROP");
                        intent.setData(uri);
                        intent.putExtra("outputX", 200);
                        intent.putExtra("outputY", 200);
                        intent.putExtra("aspectX", 1);
                        intent.putExtra("aspectY", 1);
                        intent.putExtra("scale", true);
                        intent.putExtra("return-data", true);
                        startActivityForResult(intent, REQUEST_CROP_IMAGE);
                    } catch (ActivityNotFoundException e) {
                        Toast.makeText(getActivity(), "Activity not found", Toast.LENGTH_SHORT).show();
                    }
                }
                break;
            case REQUEST_CROP_IMAGE:
                if (resultCode == Activity.RESULT_OK && data != null && data.getExtras() != null) {
                    Bitmap bitmap = data.getExtras().getParcelable("data");
                    speedDial.setIcon(WebIcon.createIcon(bitmap));
                    icon.setImageBitmap(bitmap);
                }
                break;
        }
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        try {
            mCallBack = (SpeedDialEditCallBack) getActivity();
            goBack = (GoBackController) getActivity();
        } catch (ClassCastException e) {
            throw new IllegalStateException(e);
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mCallBack = null;
        goBack = null;
    }

    interface GoBackController {
        boolean goBack();
    }
}

File: app/src/main/java/jp/hazuki/yuzubrowser/speeddial/view/SpeedDialSettingActivityEditFragment.kt
package jp.hazuki.yuzubrowser.speeddial.view

import android.app.Activity
import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import jp.hazuki.yuzubrowser.R
import jp.hazuki.yuzubrowser.download.DownloadFileProvider
import jp.hazuki.yuzubrowser.speeddial.SpeedDial
import jp.hazuki.yuzubrowser.speeddial.WebIcon
import jp.hazuki.yuzubrowser.utils.ImageUtils
import kotlinx.android.synthetic.main.fragment_edit_speeddial.*
import java.io.File

class SpeedDialSettingActivityEditFragment : Fragment() {

    private lateinit var speedDial: SpeedDial
    private var mCallBack: SpeedDialEditCallBack? = null
    private var goBack: GoBackController? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_edit_speeddial, container, false)
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        speedDial = arguments.getSerializable(DATA) as? SpeedDial ?: SpeedDial()


        superFrameLayout.setOnImeShownListener { visible -> bottomBar.visibility = if (visible) View.GONE else View.VISIBLE }

        name.setText(speedDial.title)
        url.setText(speedDial.url)

        val iconBitmap = speedDial.icon ?: WebIcon.createIcon(ImageUtils.getBitmapFromVectorDrawable(activity, R.drawable.ic_public_white_24dp))

        icon.setImageBitmap(iconBitmap.bitmap)

        use_favicon.isChecked = speedDial.isFavicon
        setIconEnable(!speedDial.isFavicon)

        use_favicon.setOnCheckedChangeListener { _, isChecked ->
            speedDial.isFavicon = isChecked
            setIconEnable(!isChecked)
        }

        icon.setOnClickListener {
            val intent = Intent(Intent.ACTION_GET_CONTENT)
            intent.type = "image/*"
            startActivityForResult(intent, REQUEST_PICK_IMAGE)
        }

        okButton.setOnClickListener {
            mCallBack?.run {
                speedDial.title = name.text.toString()
                speedDial.url = url.text.toString()
                onEdited(speedDial)
            }
        }

        cancelButton.setOnClickListener {
            goBack?.run {
                goBack()
            }
        }
    }

    private fun setIconEnable(enable: Boolean) {
        icon.isEnabled = enable
        icon.alpha = if (enable) 1.0f else 0.6f
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        when (requestCode) {
            REQUEST_PICK_IMAGE -> if (resultCode == Activity.RESULT_OK && data != null) {
                try {
                    var uri = data.data ?: return
                    if ("file" == uri.scheme) {
                        uri = DownloadFileProvider.getUriForFIle(File(uri.path))
                    }
                    val intent = Intent("com.android.camera.action.CROP").apply {
                        this.data = uri
                        putExtra("outputX", 200)
                        putExtra("outputY", 200)
                        putExtra("aspectX", 1)
                        putExtra("aspectY", 1)
                        putExtra("scale", true)
                        putExtra("return-data", true)
                    }
                    startActivityForResult(intent, REQUEST_CROP_IMAGE)
                } catch (e: ActivityNotFoundException) {
                    Toast.makeText(activity, "Activity not found", Toast.LENGTH_SHORT).show()
                }

            }
            REQUEST_CROP_IMAGE -> if (resultCode == Activity.RESULT_OK && data != null && data.extras != null) {
                val bitmap = data.extras.getParcelable<Bitmap>("data")
                speedDial.icon = WebIcon.createIconOrNull(bitmap)
                icon.setImageBitmap(bitmap)
            }
        }
    }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        try {
            mCallBack = activity as SpeedDialEditCallBack
            goBack = activity as GoBackController
        } catch (e: ClassCastException) {
            throw IllegalStateException(e)
        }
    }

    override fun onDetach() {
        super.onDetach()
        mCallBack = null
        goBack = null
    }

    internal interface GoBackController {
        fun goBack(): Boolean
    }

    companion object {
        private const val DATA = "dat"
        private const val REQUEST_PICK_IMAGE = 100
        private const val REQUEST_CROP_IMAGE = 101

        fun newInstance(speedDial: SpeedDial): Fragment {
            val fragment = SpeedDialSettingActivityEditFragment()
            val bundle = Bundle()
            bundle.putSerializable(DATA, speedDial)
            fragment.arguments = bundle
            return fragment
        }
    }
}
